clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -273,26 +273,29 @@ public RecordAppendResult append(String topic,
 
                 // check if we have an in-progress batch
                 Deque<ProducerBatch> dq = topicInfo.batches.computeIfAbsent(effectivePartition, k -> new ArrayDeque<>());
+                RecordAppendResult appendResult;
                 synchronized (dq) {
                     // After taking the lock, validate that the partition hasn't changed and retry.
                     if (topicInfo.builtInPartitioner.isPartitionChanged(partitionInfo)) {
                         log.trace("Partition {} for topic {} switched by a concurrent append, retrying",
                                 partitionInfo.partition(), topic);
                         continue;
                     }
-                    RecordAppendResult appendResult = tryAppend(timestamp, key, value, headers, callbacks, dq, nowMs);
-                    if (appendResult != null) {
+                    appendResult = tryAppend(timestamp, key, value, headers, callbacks, dq, nowMs);
+                    if (appendResult != null && !appendResult.newBatchCreated) {
```

Comment:
The built-in partitioner needs to know the size of every append, otherwise it wouldn't be able to properly switch partitions.  In particular, if records are greater than batch size and always create a new batch, it looks like the size won't be updated and partition won't get switched.  Can we add a unit test to check that partitions get switched when records are greater than the batch size?

Reply:
@artemlivshits QQ - 
[updatePartitionInfo](https://github.com/apache/kafka/blob/36bfc3a254c71498830188b207abfdc420525397/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L285) was previously getting updated only when this [non-null RecordAppendResult was returned](https://github.com/apache/kafka/blob/36bfc3a254c71498830188b207abfdc420525397/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L400) ie when 
1. Non-null ProducerBatch existed in deque. 
2. There was room to add this record in that ProducerBatch. 

With this change - [topicInfo.builtInPartitioner.updatePartitionInfo](https://github.com/apache/kafka/blob/36bfc3a254c71498830188b207abfdc420525397/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L400) will be called only for the above cases. appendResult.newBatchCreated is marked true only when ProducerBatch is found `null`. 
So invocation of `topicInfo.builtInPartitioner.updatePartitionInfo` will experience no change. 

before -> `null` ProducerBatch would result in abortOnNewBatch and finally update [builtInPartitioner](https://github.com/apache/kafka/blob/36bfc3a254c71498830188b207abfdc420525397/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L319), 
with this change, builtInPartitioner will also get updated the same place. (with appropriate appendResult.appendedBytes)

Reply:
let me try to add a unit test just in case... but iiuc the PR shouldn't change existing behaviour wrt `builtInPartitioner`

Reply:
🤔 when there's no room, records > batch-size, earlier `tryAppend` would return `null`. It will return `null` even now. 
[this condition](https://github.com/sudeshwasnik/kafka/blob/a0e9020f87b0f5014506a32f41bc431c1cbfbc88/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L285), is still doing the `null` check.